### PR TITLE
translator: Implement RDS validation

### LINKED ISSUE #11794: HTTPRoute header manipulation configuration error results in bad xDS being server to Envoy
### kgateway version

2.1.0

### Kubernetes Version

1.31.7

### Describe the bug

When you apply an `HTTPRoute` with an incorrect RequestHeaderModifier configuration, invalid xDS is being served to Envoy. This results in Envoy not being updates on any subsequent configuration change until the issue is resolved. Probably more problematic is that a restart of the data-plane will result in the removal of all routes, as the restarted Envoy data-plane can not be served correct xDS.

This is easy to reproduce with a simple httpbin deployment and the following `HTTPRoute`:

```
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: api-example-com
  namespace: default
  labels:
    app: apis
    extauth: basic
spec:
  parentRefs:
    - name: gw
      namespace: ingress-gw
  hostnames:
    - "api.example.com"
  rules:
    - matches:  
      - path:
          type: PathPrefix
          value: /
      filters:
      - type: RequestHeaderModifier
        requestHeaderModifier:
          add:
          - name: "correlation-id"
            value: "%REQ(Correlation-Id)"
      backendRefs:
      - name: httpbin
        namespace: httpbin
        port: 8000
```

Notice the `"%REQ(Correlation-Id)"`, which is missing the closing "%" character. Applying this CR results in the following errors being logged in the Gateway pod:

```
[2025-07-28 14:52:34.030][1][warning][config] [external/envoy/source/extensions/config_subscription/grpc/grpc_subscription_impl.cc:138] gRPC config for type.googleapis.com/envoy.config.route.v3.R │
│ outeConfiguration rejected: Incorrect configuration: %REQ(Correlation-Id). Couldn't find valid command at position 0
```

Note that in Gloo, this error can be caught by enabling `fullEnvoyValidation`.

# Related Issues
- https://github.com/kgateway-dev/kgateway/issues/9013
- https://github.com/solo-io/solo-projects/issues/8446

### Expected Behavior

The invalid/misconfigured resource (in this case the HTTPRoute) is either not accepted at all when applied, or its status reflects the error and the HTTPRoute is not processed and added to xDS. Under no circumstance should we ever serve bad xDS to Envoy.

### Steps to reproduce the bug

See the `HTTPRoute` in the bug description.

### Additional Environment Detail

_No response_

### Additional Context

_No response_

### LINKED ISSUE #11903: Refresh the gateway-level & listener-level standard route replacement test cases
These test cases were a placeholder to showcase that invalid policy attached at the Gateway & listener-level weren't respected in route replacement semantics. Since then, we fixed the bug those test cases were relying on, and those test cases now would fail to be persisted to the apiserver and should be updated to use the classic malformed transformation template example.

### LINKED ISSUE #11904: Ensure TrafficPolicy attachment scenarios are captured in route replacement suite
Placeholder before I forget:
 
- Gateway-level
- Listener-level for both HTTP & HTTPS listeners
- ExtensionRef-based attachment
- BackendRef has ExtensionRef-based attachment

Partially related to #11903.
Partially related to #11897.

### PR DESCRIPTION
**Pull request by timflannagan**
# Description

When kgateway is running in STRICT mode, ensure that translates routes will not produce NACKs in the data plane proxy by dropping invalid matchers, and replacing valid routes with invalid, built-in policy attached to them.

Follow-up(s):

- Optimize calls made to envoy validate mode. The 1.x implementation used an LRU cache to do this. Additionally, determine whether we could replace the call made to envoy validate mode for the matcher only scenario with improved regex validation logic, but there's some trade-off(s) here
- Decouple the custom RouteReplacementDropped condition type. This was introduced when we removed the PartiallyInvalid condition type, and before we knew we'd have an edge case where dropping (vs. replacing) routes would be needed with the route replacement functionality
- Re-organize the route replacement translator test files. The current plan is to make standard/ a subset of the strict/ test cases, and move the per-category (e.g. policy, built-in, matcher, etc.) as shared directories
- Re-evaluate the current logic where we drop routes with invalid matchers in the routeReplacementErr conditional code. This is awkward, but #11937 will need to touch this code again

Fixes #11795
Fixes #11903
Fixes #11794
Fixes #11904

# Change Type

/kind feature

# Changelog

```release-note
Extend the route replacement functionality so that when kgateway runs in STRICT mode it prevents invalid Envoy route configuration from ever reaching the proxies. It primarily covers HTTPRoute rules that either 1.) define invalid matchers (e.g. bad regular expressions) or 2.) use built-in Gateway API filters that translate into invalid Envoy xDS.
```

# Additional Notes

### CODE REVIEW

File: internal/kgateway/translator/irtranslator/validate.go [hunk 1]
```diff
@@ -27,8 +36,94 @@ var (
...
+
+	allZeroWeight := true
+	for _, cluster := range clusters {
+		if cluster.GetWeight().GetValue() > 0 {
+			allZeroWeight = false
+			break
+		}
+	}
+	if allZeroWeight {
+		*errs = append(*errs, errors.New("All backend weights are 0. At least one backendRef in the HTTPRoute rule must specify a non-zero weight"))
```
[the remark this answers is not in the document]
Comment (timflannagan) [reply internal/kgateway/translator/irtranslator/validate.go:123]: Existing code that was moved, although relevant for sure.

File: internal/kgateway/translator/gateway/gateway_translator_test.go [hunk 2]
```diff
@@ -1120,269 +1120,454 @@ func TestRouteReplacement(t *testing.T) {
 		translatortest.TestTranslation(t, ctx, inputFiles, expectedProxyFile, in.gwNN, in.assertReports, settingOpts...)
 	}
 
-	t.Run("Standard Mode - Invalid HTTPRoute Prefix Match", func(t *testing.T) {
+	t.Run("Standard/Matcher/Path Prefix Invalid", func(t *testing.T) {
```
Comment (lgadban) [inline internal/kgateway/translator/gateway/gateway_translator_test.go:1123]: will the `/` delimited naming convention make it difficult to run specific tests via `-test.run` ?

File: internal/kgateway/translator/irtranslator/route.go [hunk 3]
```diff
@@ -210,7 +220,20 @@ func (h *httpRouteConfigurationTranslator) envoyRoutes(
...
 
-		// If routeAcceptanceErr is set, report Accepted=False with Reason=RouteRuleDropped
+		// For invalid matchers, we drop the route entirely instead of replacing it with a synthetic matcher.
+		// TODO(tim): Extract this logic outside of the routeReplacementErr check.
+		if errors.Is(routeReplacementErr, ErrInvalidMatcher) {
+			h.logger.Info("invalid matcher", "error", routeReplacementErr)
+			routeReport.SetCondition(reportssdk.RouteCondition{
+				Type:    gwv1.RouteConditionAccepted,
+				Status:  metav1.ConditionFalse,
+				Reason:  gwv1.RouteConditionReason(reportssdk.RouteRuleDroppedReason),
```
Comment (lgadban) [inline internal/kgateway/translator/irtranslator/route.go:230]: We had previously using the dropped reason here while changing the reason in an actual replacement scenario to not be dropped and indicate it was replaced. Not doing in this PR seems fine but I really think we still need to take this on soon. Wdyt?

File: internal/kgateway/translator/irtranslator/route.go [hunk 4]
```diff
@@ -194,6 +199,11 @@ func (h *httpRouteConfigurationTranslator) envoyRoutes(
 		routeProcessingErr = errors.New("no action specified")
 	}
 
+	// If there are no errors, validate the route will not be rejected by the xDS server.
+	if routeProcessingErr == nil {
+		routeProcessingErr = validateRoute(ctx, out, h.validator, h.routeReplacementMode)
```
Comment (lgadban) [inline internal/kgateway/translator/irtranslator/route.go:204]: shouldn't we set this to be `routeAcceptanceErr` ?
seems like thus far `routeProcessingErr` is really just a case of an err from route plugins returning an err.
maybe we should rename it to better reflect that as well

File: internal/kgateway/translator/irtranslator/validate.go [hunk 5]
```diff
@@ -27,8 +36,94 @@ var (
...
+// validation determines if the route should be replaced with a direct response due to other
+// configuration issues.
+//
+// This two-tiered approach is necessary because Envoy validate mode output does not provide
+// enough information to determine if the error is due to an invalid matcher (drop route)
+// or other route configuration issues (replace with direct response).
+//
+// TODO(tim): optimize the calls made to envoy validate mode with this approach, or add better
+// regex validation throughout the route translator to handle the invalid matcher scenario w/o
+// shelling out to envoy validate mode.
```
Comment (lgadban) [inline internal/kgateway/translator/irtranslator/validate.go:64]: unfortunate but makes sense 👍

File: internal/kgateway/translator/gateway/gateway_translator_test.go (hunk 2 above)
Comment (timflannagan) [reply internal/kgateway/translator/gateway/gateway_translator_test.go:1123]: I made this change exactly for that reason (i.e. making it easier to run via `go test -run ...` now that this suite was migrated away from ginkgo. Ex: `go test -v -failfast -run "TestRouteReplacement/Standard/Matcher" ./internal/kgateway/translator/gateway/` to run a specific mode's sub-category, or just `TestRouteReplacement/Standard` if you want to only run a specific mode's route replacement tests.

File: internal/kgateway/translator/irtranslator/route.go (hunk 3 above)
Comment (timflannagan) [reply internal/kgateway/translator/irtranslator/route.go:230]: Yep, good call. I have this tracked locally, but forgot to bubble up that as a follow-up bullet point, which as aside, I just added to the description before this review was posted. Agreed on doing this as soon as possible. Just wanted to limit the changes here given the LOC diff.

File: internal/kgateway/translator/irtranslator/validate.go (hunk 5 above)
Comment (timflannagan) [reply internal/kgateway/translator/irtranslator/validate.go:64]: I think the LRU caching and current short circuiting logic will help out, but open to suggestions if you have any. Unfortunately, we'd either need to invest in granular output for envoy validate mode failures so we can parse/distinguish between invalid matchers vs. invalid typed filter config, or we rely on regex validation to handle the invalid matcher scenario, and risk the low N chance that invalid RDS config is set to proxies.

File: internal/kgateway/translator/irtranslator/route.go (hunk 4 above)
Comment (timflannagan) [reply internal/kgateway/translator/irtranslator/route.go:204]: Yeah.. I'm not sure what's best right now. FWIW, this is still consistent with what main is doing:

https://github.com/kgateway-dev/kgateway/blob/1b3346d412af11e48b0ff91a9721c04d76490731/internal/kgateway/translator/irtranslator/route.go#L173-L175

The reason I chose this approach vs. extending routeAcceptanceErr is that the latter indexes into the input IR for it's RouteAcceptanceError (or RouteReplacementError in the context of the routeReplacementErr), and I wanted to keep that distinction, especially since there's no equivalent RouteDroppedErr IR type. WYDT? I can easily move the logic around dropping the route for invalid matchers outside of the routeReplacementErr conditional:

```diff
+       // If the route is invalid, we drop the route entirely instead of replacing it with a synthetic matcher.
+       if routeAcceptanceErr != nil && errors.Is(routeAcceptanceErr, ErrInvalidMatcher) {
+               h.logger.Info("invalid matcher", "error", routeAcceptanceErr)
+               routeReport.SetCondition(reportssdk.RouteCondition{
+                       Type:    gwv1.RouteConditionAccepted,
+                       Status:  metav1.ConditionFalse,
+                       Reason:  gwv1.RouteConditionReason(reportssdk.RouteRuleDroppedReason),
+                       Message: fmt.Sprintf("Dropped Rule (%d): %v", in.MatchIndex, routeAcceptanceErr),
+               })
+               return nil
+       }
+
        // If routeReplacementErr is set, we need to replace the route with a direct response
        if routeReplacementErr != nil {
                h.logger.Debug("invalid route", "error", routeReplacementErr)

-               // For invalid matchers, we drop the route entirely instead of replacing it with a synthetic matcher.
-               // TODO(tim): Extract this logic outside of the routeReplacementErr check.
-               if errors.Is(routeReplacementErr, ErrInvalidMatcher) {
-                       h.logger.Info("invalid matcher", "error", routeReplacementErr)
-                       routeReport.SetCondition(reportssdk.RouteCondition{
-                               Type:    gwv1.RouteConditionAccepted,
-                               Status:  metav1.ConditionFalse,
-                               Reason:  gwv1.RouteConditionReason(reportssdk.RouteRuleDroppedReason),
-                               Message: fmt.Sprintf("Dropped Rule (%d): %v", in.MatchIndex, routeReplacementErr),
-                       })
-                       return nil
-               }
```

File: internal/kgateway/translator/irtranslator/route.go (hunk 3 above)
Comment (lgadban) [reply internal/kgateway/translator/irtranslator/route.go:230]: So the plan will be to address this when tackling the higher-level attachment validation? If so that makes sense, let's just make sure it's included in scope wherever we are tracking that.

File: internal/kgateway/translator/gateway/gateway_translator_test.go (hunk 2 above)
Comment (lgadban) [reply internal/kgateway/translator/gateway/gateway_translator_test.go:1123]: Ah ok, so we don't need to e.g. escape the `/` to pass it as a test pattern? That's great

File: internal/kgateway/translator/irtranslator/route.go (hunk 3 above)
Comment (timflannagan) [reply internal/kgateway/translator/irtranslator/route.go:230]: My plan was to tackle that separate from that work. Likely as a distinct issue / standalone follow-up. Same with all those bullet points mentioned above that don't currently have explicit trackers. I was going to create them once this merges.

File: internal/kgateway/translator/irtranslator/route.go (hunk 4 above)
Comment (lgadban) [reply internal/kgateway/translator/irtranslator/route.go:204]: >WYDT? I can easily move the logic around dropping the route for invalid matchers outside of the routeReplacementErr conditional

I do think my biggest hang up is that it we are checking the routeReplacementErr for an error that is tied more closely to the Acceptance of the route, thus using `routeAcceptanceErr` seems more intuitive.

If we can easily reorder this I think that would be sufficient IMO but I don't feel like this is necessarily a blocker either.

Comment (timflannagan) [reply internal/kgateway/translator/irtranslator/route.go:204]: Right, I think we're on the same page that it 1.) reads awkward at best, and 2.) introduces some minor form of cognitive load. This is what I was mentioning in the PR description as something we need to iterate on going forward. My preference is we see what the route translator looks like after #11937 has consensus, gets implemented, etc.

File: internal/kgateway/translator/gateway/gateway_translator_test.go (hunk 2 above)
Comment (timflannagan) [reply internal/kgateway/translator/gateway/gateway_translator_test.go:1123]: Exactly - for running an individual sub-test, you'd hit the "run test" over the t.Run block depending on the IDE (e.g. vscode, or forks of vscode like cursor support this today, which was the intention outlined in the refactor PR). For an individual test though, spaces get escaped (translates?) into underscores, so a bit awkward to directly run via a -run flag.

File: internal/kgateway/translator/irtranslator/route.go (hunk 4 above)
Comment (lgadban) [reply internal/kgateway/translator/irtranslator/route.go:204]: will be fixed in a follow up

**Review by lgadban — APPROVED** [review]
